src/SearchField/SearchFieldAdvanced.jsx

```diff
@@ -0,0 +1,172 @@
+/* eslint-disable max-len */
+import React, { useRef, createContext, useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSearch, faTimes } from '@fortawesome/free-solid-svg-icons';
+
+import newId from '../utils/newId';
+
+export const SearchFieldContext = createContext();
+
+const SearchFieldAdvanced = (props) => {
+  const {
+    children,
+    className,
+    screenReaderText,
+    icons,
+    onSubmit,
+    onClear,
+    onChange,
+    onBlur,
+    onFocus,
+    value: initialValue,
+  } = props;
+
+  const [hasFocus, setHasFocus] = useState(false);
+  const [value, setValue] = useState(initialValue);
+
+  const isInitialMount = useRef(true);
+  const inputId = useRef(`${newId('pgn-searchfield-input-')}`);
+  const inputRef = useRef();
+  const submitButtonRef = useRef();
+
+  useEffect(() => {
+    if (isInitialMount.current) {
+      isInitialMount.current = false;
+    } else {
+      onChange(value);
+    }
+  }, [value]);
+
+  const handleSubmit = (event) => {
+    event.preventDefault();
+    onSubmit(value);
+    if (submitButtonRef && submitButtonRef.current) {
+      submitButtonRef.current.focus();
+    }
+  };
+
+  const handleClear = () => {
+    setValue('');
+    onClear();
+    if (inputRef && inputRef.current) {
+      inputRef.current.focus();
+    }
+  };
+
+  const handleFocus = (event) => {
+    setHasFocus(true);
+    onFocus(event);
+  };
+
+  const handleBlur = (event) => {
+    setHasFocus(false);
+    onBlur(event);
+  };
+
+  const handleChange = (event) => {
+    setValue(event.target.value);
+  };
+
+  return (
+    <div
+      className={classNames(
+        'pgn-searchfield',
+        'd-flex border rounded',
+        { 'has-focus': hasFocus },
+        className,
+      )}
+    >
+      <form
```

Comment:
Do you have any thoughts about whether it's good to have the form in here? I'm imagining a search with filters/checkboxes type situation where those other inputs wouldn't be able to share the same form element as the search field itself.

Reply:
Good question 🤔

For a11y, a form that encompasses a search input field should have the `role="search"` attribute. Given that, I feel that it makes sense to include the form element within the component to guarantee a11y when using this component.

If there were a filters/checkboxes type of situation alongside the search field to further narrow down results, I'm not sure it's safe to assume those controls would be within the same form as the search field input (e.g., with the same `onSubmit` handler, etc.).

The implementation of this refactor also relies on the input field and submit/clear buttons being children of a form in order to take advantage of native form submission and resetting behaviors. I'd be worried we wouldn't be able to enforce that a form wraps the `SearchField` component if we don't provide one ourselves.

FWIW, I know Algolia's `SearchBox` component includes a form element as well.

Happy to hear any additional thoughts you might have on this, of course.

Reply:
Makes sense to me, no objections.